teal/packer.tl

```diff
@@ -0,0 +1,130 @@
+local M = {}
+
+local function apply_config(plugin: Plugin)
+  if plugin.config and plugin.loaded then
+    local c = plugin.config
+    if c is function() then
+      c()
+    else
+      loadstring(c, plugin.name..'.config()')()
+    end
+  end
+end
+
+local function loader(lplugins: {Plugin})
+  for _, plugin in ipairs(lplugins) do
+    if not plugin.loaded then
+      -- Set the plugin as loaded before config is run in case something in the
+      -- config tries to load this same plugin again
+      plugin.loaded = true
+      vim.cmd.packadd(plugin.name)
+      apply_config(plugin)
+    end
+  end
+end
+
+-- Completion user plugins
+-- Intended to provide completion for PackerUpdate/Sync/Install command
+local function plugin_complete(lead: string, _:string): {string}
+  local completion_list = vim.tbl_filter(function(name: string): boolean
+    return vim.startswith(name, lead)
+  end, vim.tbl_keys(_G.packer_plugins))
+  table.sort(completion_list)
+  return completion_list
+end
+
+local function load_plugin_configs()
+  local Handlers = require('packer.handlers')
+
+  local cond_plugins: {Handlers.Type:{string:Plugin}} = {
+    cmd   = {},
+    keys  = {},
+    ft    = {},
+    event = {},
+  }
+
+  local uncond_plugins = {}
+
+  local conds = {'cmd', 'keys', 'ft', 'event'} as {Handlers.Type}
+
+  for name, plugin in pairs(_G.packer_plugins) do
```

Comment:
Yeah sure. Personally I found `_G.packer_plugins` useful as a user for basic debug and it kind of serves a similar role as `_G.package.loaded`. Obviously, populating the global scope should be done lightly, but I think a package manager has some good authority of the global scope.

> It seems cleaner to make it a module-level table. That way, extensions to packer can still access it through require('packer').plugins (or similar), but we don't have to pollute the global state.

Sure, this seems reasonable.

Reply:
Actually, putiing plugins in the main module (`packer.tl`) creates a circular dependency in places since a few modules use `_G.packer_plugins`. Probably best to put in `packer/plugin.tl` and provide an accessor in `packer.tl`.